### ISSUE TITLE
fix: EA injects large control defaults into every element → HTTP 413 on save

### ISSUE THREAD

**Issue by ZebaAfiaShama** [WPDevelopers/essential-addons-for-elementor-lite#844]
**Fix branch:** `fix-elementor-data-bloat-413` (base `dev-pr`)

---

##Card: https://projects.startise.com/wp-admin/admin.php?page=fluent-boards#/boards/30/tasks/83800-Bug-Fix-%7C-Essential-Addons-%7C-E

## Problem
Saving an Elementor page returns **HTTP 413** on hosts with strict POST limits whenever Essential Addons is active, **even on pages with zero EA widgets**. Stored `_elementor_data` balloons (client: 123 KB → 1.12 MB).

## Root cause (Lite portion)
`Image_Masking.php` and `Vertical_Text_Orientation.php` register controls on every common element (widget / section / column / container) with large non-empty `default` values:

- `eael_image_masking_custom_clip_path` (+ `_hover`) — clip-path polygon default
- `eael_vto_writing_gradient_color_repeater` — two-stop gradient repeater default

Elementor cannot strip these on save (repeater rows get a runtime `_id`; the scalar default diverges), so they serialize into `_elementor_data` and every save POST for every element regardless of feature use.

## Fix
Empty each control default; apply the real default at render time — clip-path fallback inline, gradient via `eael_vto_default_gradient_colors()`. No behavior change.

## Verification (sandbox: Lite 6.6.10 + Pro 7.0.1, Elementor 4.1.5)
8 core widgets, 0 EA widgets — stored `_elementor_data` after a real editor save: **~24 KB → 831 bytes, 0 `eael_` keys**. No PHP errors; page renders HTTP 200.

Companion Pro PR: WPDevelopers/essential-addons-elementor#289 (same pattern for Custom Cursor, Extender/Image-Masking-morphing, Conditional Display, Content Protection, Tooltip).

---

<details>
<summary>Full diff</summary>

```diff
diff --git a/includes/Extensions/Image_Masking.php b/includes/Extensions/Image_Masking.php
index f9b2ee917..14a5cb8d4 100644
--- a/includes/Extensions/Image_Masking.php
+++ b/includes/Extensions/Image_Masking.php
@@ -178,7 +178,9 @@ class Image_Masking {
                 'ai' => [
 					'active' => false,
 				],
-                'default'     => 'clip-path: polygon(50% 0%, 80% 10%, 100% 35%, 100% 70%, 80% 90%, 50% 100%, 20% 90%, 0% 70%, 0% 35%, 20% 10%);',
+                // Default kept empty so this is not serialized into every element; the starter
+                // shape is applied at render time when custom clip path is enabled but left blank.
+                'default'     => '',
                 'placeholder' => 'clip-path: polygon(50% 0%, 0% 100%, 100% 100%);',
                 'description' => __( 'Enter your custom clip path value. You can use <a href = "https://bennettfeely.com/clippy/" target = "_blank">Clippy</a> to generate your custom clip path.', 'essential-addons-for-elementor-lite' ),
                 'condition'   => array_merge( $condition, [ 'eael_image_masking_enable_custom_clip_path' . $tab => 'yes' ] ),
@@ -563,7 +565,7 @@ class Image_Masking {
 			if( 'clip' === $type ){
                 $clip_path_value = '';
                 if( 'yes' === $settings['eael_image_masking_enable_custom_clip_path'] ){
-                    $clip_path_value = $settings['eael_image_masking_custom_clip_path'];
+                    $clip_path_value = ! empty( $settings['eael_image_masking_custom_clip_path'] ) ? $settings['eael_image_masking_custom_clip_path'] : 'clip-path: polygon(50% 0%, 80% 10%, 100% 35%, 100% 70%, 80% 90%, 50% 100%, 20% 90%, 0% 70%, 0% 35%, 20% 10%);';
                     $clip_path_value = str_replace( 'clip-path: ', '', $clip_path_value );
                     $clip_path_value = self::sanitize_css_fragment( $clip_path_value );
                 } else {
@@ -577,7 +579,7 @@ class Image_Masking {
                 if( 'yes' === $settings['eael_image_masking_hover_effect'] ){
                     $hover_clip_path_value = '';
                     if( 'yes' === $settings['eael_image_masking_enable_custom_clip_path_hover'] ){
-                        $hover_clip_path_value = $settings['eael_image_masking_custom_clip_path_hover'];
+                        $hover_clip_path_value = ! empty( $settings['eael_image_masking_custom_clip_path_hover'] ) ? $settings['eael_image_masking_custom_clip_path_hover'] : 'clip-path: polygon(50% 0%, 80% 10%, 100% 35%, 100% 70%, 80% 90%, 50% 100%, 20% 90%, 0% 70%, 0% 35%, 20% 10%);';
                         $hover_clip_path_value = str_replace( 'clip-path: ', '', $hover_clip_path_value );
                         $hover_clip_path_value = self::sanitize_css_fragment( $hover_clip_path_value );
                     } else {
diff --git a/includes/Extensions/Vertical_Text_Orientation.php b/includes/Extensions/Vertical_Text_Orientation.php
index 6c1b3a587..af51bc06b 100644
--- a/includes/Extensions/Vertical_Text_Orientation.php
+++ b/includes/Extensions/Vertical_Text_Orientation.php
@@ -309,22 +309,10 @@ class Vertical_Text_Orientation {
 				'label'   => esc_html__( 'Choose Gradient Colors', 'essential-addons-for-elementor-lite' ),
 				'type'    => Controls_Manager::REPEATER,
 				'fields'  => $repeater->get_controls(),
-				'default' => [
-					[
-						'eael_vto_writing_gradient_color' => '#7C62FF',
-                        'eael_vto_writing_gradient_color_location' => [
-                            'unit' => '%',
-                            'size' => 50,
-                        ],
-					],
-					[
-						'eael_vto_writing_gradient_color' => '#FF6464',
-                        'eael_vto_writing_gradient_color_location' => [
-                            'unit' => '%',
-                            'size' => 90,
-                        ],
-					],
-				],
+				// Default kept empty: a non-empty repeater default is serialized (with per-row _id)
+				// into every element and never stripped by Elementor. The starter gradient is applied
+				// at render time via eael_vto_default_gradient_colors() when the feature is enabled.
+				'default' => [],
 				'title_field' => '{{{ eael_vto_writing_gradient_color }}}',
                 'condition' => [
 					'eael_vertical_text_orientation_switch' => 'yes',
@@ -534,11 +522,28 @@ class Vertical_Text_Orientation {
 		$element->end_controls_section();
 	}
 
+	/**
+	 * Starter gradient colours, applied at render time when the user has defined none.
+	 * Kept out of the control default so they are not serialized into every element.
+	 */
+	private function eael_vto_default_gradient_colors() {
+		return [
+			[ 'eael_vto_writing_gradient_color' => '#7C62FF', 'eael_vto_writing_gradient_color_location' => [ 'unit' => '%', 'size' => 50 ] ],
+			[ 'eael_vto_writing_gradient_color' => '#FF6464', 'eael_vto_writing_gradient_color_location' => [ 'unit' => '%', 'size' => 90 ] ],
+		];
+	}
+
 	public function before_render( $element ) {
 		$settings = $element->get_settings_for_display();
-        if ( !empty( $settings['eael_vto_writing_gradient_color_repeater'] ) && is_array( $settings['eael_vto_writing_gradient_color_repeater'] ) ) {
+        $gradient_repeater = $settings['eael_vto_writing_gradient_color_repeater'] ?? [];
+        if ( empty( $gradient_repeater )
+            && 'yes' === ( $settings['eael_vertical_text_orientation_switch'] ?? '' )
+            && 'gradient' === ( $settings['eael_vto_writing_styling_type'] ?? '' ) ) {
+            $gradient_repeater = $this->eael_vto_default_gradient_colors();
+        }
+        if ( !empty( $gradient_repeater ) && is_array( $gradient_repeater ) ) {
             $gradient_colors = [];
-            foreach ( $settings['eael_vto_writing_gradient_color_repeater'] as $value ) {
+            foreach ( $gradient_repeater as $value ) {
                 $gradient_colors[] = [
                     'color' => $value['eael_vto_writing_gradient_color'],
                     'location' => $value['eael_vto_writing_gradient_color_location']['size'] . $value['eael_vto_writing_gradient_color_location']['unit'],

```
</details>

**Comment by ZebaAfiaShama** [comment]
Branch pushed to this repository: [`fix-elementor-data-bloat-413`](https://github.com/WPDevelopers/essential-addons-for-elementor-lite/tree/fix-elementor-data-bloat-413) — [diff against `dev-pr`](https://github.com/WPDevelopers/essential-addons-for-elementor-lite/compare/dev-pr...fix-elementor-data-bloat-413). Ready for dev/QA review.

**Comment by coderskamrul** [comment]
How to test
1. Reset a page: new Elementor page, add 8 core widgets (heading/text/image/button…), no EA widgets. Save.

2. Measure:
SELECT LENGTH(meta_value), (LENGTH(meta_value)-LENGTH(REPLACE(meta_value,'eael_','')))/5
FROM wp_postmeta WHERE post_id=<ID> AND meta_key='_elementor_data';
Expect ~800 bytes / 0 eael_ keys. Before the fix: ~24 KB / ~96 keys.

3. Save payload: DevTools → Network → hit Update → find admin-ajax.php (action=elementor_ajax) → Request payload size. Expect ~3.7 KB, not ~36 KB.

4. Behavior parity — each feature must render identically with its field left blank:
Custom Cursor → type SVG Code, leave blank → red pin cursor still shows (editor + frontend).
Custom Cursor → Trail → frostsparkles, blank emoji field → ❄️ still trails.
Image Masking → Morphing → clip-path, don't touch Shapes → 4-shape morph still animates.
Conditional Display → enable, don't add a logic row → still behaves as "logged in" rule.
Content Protection → password type, blank text fields → placeholder="Enter Password", value="Submit", and wrong-password shows "Password does not match."
Advanced Tooltip → enable, blank Content → "I am a tooltip" still shows.

5. Regression check: WP_DEBUG=true, load a page with each extension active — no notices/deprecations.
